docs/src/DeveloperDocumentation/CachedFunctions.md

```diff
@@ -0,0 +1,367 @@
+# Caching
+
+Several functions in OSCAR cache the output for different reasons and in different ways.
+We describe this behaviour in this section.
+
+## Caching the results of complex computations
+
+Many functions which are expected to have a long runtime in general (say, a function `slow_function`) store the result of the computation on the input (say, `some_input`).
+Subsequent calls of `slow_function(some_input)` will then return the identical result and be (nearly) instantaneous.
+For example, if `K` is a number field, the output of `maximal_order(K)` will be stored in `K`.
+Another prominent example would be the function `groebner_basis(::MPolyIdeal)` which stores the computed basis in the given ideal (together with the used monomial ordering).
+This kind of caching is performed either via using a field in the input (say, `some_input.slow_result`) or using attributes.
+For example, `groebner_basis(I)` stores the result in the dictionary `I.gb`, while `maximal_order(K)` uses attributes to store the maximal order.
+The only reason to disable storing hard-to-compute results should be for debugging purposes.
+We hence do not provide an option to disable this kind of caching in general.
+
+## Caching of sub or quotient structures
+
+Some structures allow caching of sub or quotient structures in a global 'lattice' object.
+At the time of this writing, this is only implemented for finitely generated abelian groups (`FinGenAbGroup`).
+For these, given a group `G` the call `H, _ = sub(G, [...])` stores `G` and `H` in a global subgroup lattice which, for example, enables the user to enter arithmetic expression like `G[1] + H[1]` without the need to explicitly map elements from one group into the other.
+The user may disable the storage of a group in the lattice with the argument `add_to_lattice` of the functions `sub`, `quo`, etc.
+
+## Caching of parent objects
+
+Another kind of caching concerns 'parent objects', that is, functions that create an ambient ring or field, in which subsequent calculations are performed.
+Many such functions cache the constructed object in a global dictionary.
+Examples are the functions `polynomial_ring` or `number_field` where creating the same ring or field twice will return the identical object.
+```julia
+julia> Qx, x = QQ[:x]
+(Univariate polynomial ring in x over QQ, x)
+
+julia> QQx, xx = QQ[:x]
+(Univariate polynomial ring in x over QQ, x)
+
+julia> Qx === QQx
+true
+
+julia> x == xx
+true
+```
+This is mainly intended for interactive use where this behaviour may be convenient.
+However, the parent objects constructed in this way cannot be deleted (that is, garbage collected) as they are stored in the dictionary.
```

Comment:
I believe this is wrong, at least this is what I understand from the WeakValueDict docstring in AbstractAlgebra

> `WeakValueDict()` constructs a hash table where the values are weak
references to objects which may be garbage collected even when
used as values in a hash table.

Reply:
This is how I understood it from the OSCAR book.
But yes, apparently garbage collection works:
```
julia> Qx, x = QQ[:x]
(Univariate polynomial ring in x over QQ, x)

julia> Oscar.Nemo.FmpqPolyID
AbstractAlgebra.WeakValueDict{Symbol, QQPolyRing}(...):
  :x => Univariate polynomial ring in x over QQ

julia> Qx = 0
0

julia> x = 0
0

julia> GC.gc()

julia> Oscar.Nemo.FmpqPolyID
AbstractAlgebra.WeakValueDict{Symbol, QQPolyRing}()
```
Conclusion: I'm not qualified either to write this documentation.